### PR TITLE
[FLAX] Add dtype to embedding for bert/bart/opt/t5

### PR DESCRIPTION
**Pull request by merrymercy**
## What does this PR do?

This PR is the follow-up of #18462. It adds dtype to `nn.Embed` for more common Flax models, including bert, bart, opt, t5, and their copies.
This dtype is necessary for mixed precision training.

## Who can review?
@patrickvonplaten, @LysandreJik, @sanchit-gandhi

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
cc @sanchit-gandhi @younesbelkada @ArthurZucker here

**Review by younesbelkada — APPROVED** [review]
LGTM as we do that for BLOOM Flax too: https://github.com/huggingface/transformers/pull/18022/files#r959516589 
this is important for large embeddings (BLOOM has a vocab size of ~25000) this is very important to match the logits between PT and Flax implementation
Does slow tests pass with this fix? (especially for Flax T5)
Would wait for @ArthurZucker and/or @sanchit-gandhi for final approval!

**Review by sanchit-gandhi — APPROVED** [review]
LGTM, thanks @merrymercy! Worth checking that the slow tests pass as @younesbelkada has mentioned 🤗

EDIT: the aforementioned test doesn't do any dytpe upcasting/downcasting, so I'd expect it to pass. What would be good would be adding a test where we do a forward pass with dtype=bfloat16.

**Comment by merrymercy** [comment]
Can we merge this @ArthurZucker ? How to check whether the slow tests pass?

**Comment by younesbelkada** [comment]
Thanks @merrymercy !
You can run slow tests for `t5` for example by running: `RUN_SLOW=1 pytest tests/models/t5/test_modeling_flax_t5.py` - regarding what has been suggested by @sanchit-gandhi , you can just add a new test that initializes a model let's say in `bf16` and tests if the generated sequence is the same than the one which is expected:
```
    @slow
    def test_small_generation_bf16(self):
        model = FlaxT5ForConditionalGeneration.from_pretrained("t5-small", dtype=jnp.bfloat16)

        self.assertTrue(model.params["shared"]["embedding"].dtype == jnp.bfloat16)
        model.config.max_length = 8
        model.config.num_beams = 1
        model.config.do_sample = False
        tokenizer = T5Tokenizer.from_pretrained("t5-small")

        input_ids = tokenizer("summarize: Hello there", return_tensors="np").input_ids

        sequences = model.generate(input_ids).sequences

        output_str = tokenizer.batch_decode(sequences, skip_special_tokens=True)[0]
        self.assertTrue(output_str == "Hello there!")

```

**Comment by merrymercy** [comment]
A bf16 test case is added as suggested by @younesbelkada . I checked that slow tests are also passed because most slow tests are in fp32 and this PR does not change the behavior of any fp32 tests.

However, I do notice some slow tests fail on my machine (V100, jax=0.3.25, flax=0.6.2), even with the transformers main branch and transformers v4.24.0, so I think this is not related to my PR. I can confirm that all tests passed on the main branch can still be passed after my PR.

Since we get two approvals and the test case is added, can we merge it now?

**Comment by sgugger** [comment]
Thanks again for your contribution!

**Pull request merged**
